File: server/src/models/Room.ts
import WebSocket from "ws";
import {
    CustomWebSocket,
    DisconnectedUserInfo,
    ParticipantRecord,
    ParticipantView,
    ServerPayload
} from "../types/room.js";

export class Room {
    public readonly id: string;
    public participants: Map<string, ParticipantRecord> = new Map();
    public disconnectedUsers: Map<string, DisconnectedUserInfo> = new Map();
    public readonly createdAt: number;

    constructor(id: string, createdAt: number) {
        this.id = id;
        this.createdAt = createdAt;
    }

    public broadcast(payload: ServerPayload): void {
        const msg = JSON.stringify(payload);
        for (const participant of this.participants.values()) {
            const client = participant.socket;
            if (client && client.readyState === WebSocket.OPEN) {
                client.send(msg);
            }
        }
    }

    public sendToUser(ws: WebSocket | CustomWebSocket, payload: ServerPayload): void {
        if (ws.readyState === WebSocket.OPEN) {
            ws.send(JSON.stringify(payload));
        }
    }

    public addParticipant(participant: ParticipantRecord, socket: CustomWebSocket): void {
        socket.roomId = this.id;
        socket.participantId = participant.participantId;
        socket.username = participant.username;

        participant.connected = true;
        participant.lastSeenAt = Date.now();
        participant.socket = socket;
        this.participants.set(participant.participantId, participant);
    }

    public removeParticipant(participantId: string): ParticipantRecord | null {
        const participant = this.participants.get(participantId);
        if (!participant) return null;

        this.participants.delete(participantId);
        participant.connected = false;
        participant.lastSeenAt = Date.now();
        participant.socket = null;
        return participant;
    }

    public hasUsername(username: string): boolean {
        for (const participant of this.participants.values()) {
            if (participant.username === username) return true;
        }
        for (const disconnected of this.disconnectedUsers.values()) {
            if (disconnected.participant.username === username) return true;
        }
        return false;
    }

    public getParticipantBySocket(socket: CustomWebSocket): ParticipantRecord | null {
        if (!socket.participantId) return null;
        return this.participants.get(socket.participantId) || null;
    }

    public getParticipantView(): ParticipantView[] {
        return [...this.participants.values()]
            .map(({ participantId, username, connected, lastSeenAt }) => ({
                participantId,
                username,
                connected,
                lastSeenAt
            }))
            .sort((a, b) => a.username.localeCompare(b.username));
    }

    public get userCount(): number {
        return this.participants.size;
    }

    public get isEmpty(): boolean {
        return this.participants.size === 0 && this.disconnectedUsers.size === 0;
    }

    public clearAllReconnectTimers(): void {
        for (const info of this.disconnectedUsers.values()) {
            clearTimeout(info.timer);
        }
    }
}

File: backend-go/room.go
package main

import (
	"slices"
	"sync"
	"time"
)

type Room struct {
	ID                string
	CreatedAt         int64
	mu                sync.RWMutex
	participants      map[string]*ParticipantRecord
	disconnectedUsers map[string]*DisconnectedUserInfo
}

func NewRoom(id string, createdAt int64) *Room {
	return &Room{
		ID:                id,
		CreatedAt:         createdAt,
		participants:      make(map[string]*ParticipantRecord),
		disconnectedUsers: make(map[string]*DisconnectedUserInfo),
	}
}

func (r *Room) addParticipant(participant *ParticipantRecord, socket *WSConn) {
	now := time.Now().UnixMilli()
	socket.SetParticipant(r.ID, participant.ParticipantID, participant.Username)
	participant.Connected = true
	participant.LastSeenAt = now
	participant.Socket = socket
	r.participants[participant.ParticipantID] = participant
}

func (r *Room) removeParticipant(participantID string) *ParticipantRecord {
	participant, ok := r.participants[participantID]
	if !ok {
		return nil
	}

	delete(r.participants, participantID)
	participant.Connected = false
	participant.LastSeenAt = time.Now().UnixMilli()
	participant.Socket = nil
	return participant
}

func (r *Room) hasUsername(username string) bool {
	for _, participant := range r.participants {
		if participant.Username == username {
			return true
		}
	}
	for _, disconnected := range r.disconnectedUsers {
		if disconnected.Participant.Username == username {
			return true
		}
	}
	return false
}

func (r *Room) getParticipantView() []ParticipantView {
	views := make([]ParticipantView, 0, len(r.participants))
	for _, participant := range r.participants {
		views = append(views, ParticipantView{
			ParticipantID: participant.ParticipantID,
			Username:      participant.Username,
			Connected:     participant.Connected,
			LastSeenAt:    participant.LastSeenAt,
		})
	}
	slices.SortFunc(views, func(a, b ParticipantView) int {
		switch {
		case a.Username < b.Username:
			return -1
		case a.Username > b.Username:
			return 1
		default:
			return 0
		}
	})
	return views
}

func (r *Room) userCount() int {
	return len(r.participants)
}

func (r *Room) isEmpty() bool {
	return len(r.participants) == 0 && len(r.disconnectedUsers) == 0
}

func (r *Room) clearAllReconnectTimers() {
	for _, info := range r.disconnectedUsers {
		if info.Timer != nil {
			info.Timer.Stop()
		}
	}
}
